### PR TITLE
docker: 1.10.3 -> 1.12.1

### PR DESCRIPTION
**Pull request by offlinehacker**
###### Motivation for this change

Docker we have is very old. This build is also much better as previous, it provides man files, has dependency clean-up and removes uneeded references to go compiler. I've tested this with and without nixos and works without any issues. It would be nice to have this merged in 16.09 @domenkozar 
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/virtualization/runc/default.nix [hunk 1]
```diff
@@ -0,0 +1,64 @@
+{ stdenv, fetchFromGitHub, go-md2man
+, go, pkgconfig, libapparmor, apparmor-parser, libseccomp }:
+
+with stdenv.lib;
+
+stdenv.mkDerivation rec {
+  name = "runc-${rev}";
+  rev = "cc29e3dd"; # this version is used by docker
```
Comment (bjornfor) [inline pkgs/applications/virtualization/runc/default.nix]: So we cannot use the latest release? If the current snapshot is used by docker, that does help, but it's bad if we cannot use the release. (IMHO.)

Package versions should be monotonically increasing. A git hash is not (it jumps all over the place). Please substitute ```${rev}``` with the corresponding commit date (YYYY-MM-DD).

File: pkgs/development/tools/misc/md2man/default.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{ stdenv, lib, buildGoPackage, go, fetchFromGitHub }:
+
+with lib;
+
+buildGoPackage rec {
+  name = "go-md2man-${version}";
+  version = "v1.0.6";
```
Comment (bjornfor) [inline pkgs/development/tools/misc/md2man/default.nix]: Version must start with a digit, or else Nix' version parsing breaks:

```
ix-repl> builtins.parseDrvName "go-md2man-v1.0.6"
{ name = "go-md2man-v1.0.6"; version = ""; }
```

File: nixos/modules/virtualisation/docker.nix [hunk 3]
```diff
@@ -40,13 +40,24 @@ in
       };
     storageDriver =
       mkOption {
-        type = types.enum ["aufs" "btrfs" "devicemapper" "overlay" "zfs"];
-        default = "devicemapper";
+        type = types.nullOr (types.enum ["aufs" "btrfs" "devicemapper" "overlay" "zfs"]);
+        default = null;
         description =
           ''
             This option determines which Docker storage driver to use.
```
Comment (bjornfor) [inline nixos/modules/virtualisation/docker.nix]: Please add to description what happens with the 'null' value. I.e., what is the (new) default storage driver?

**Comment by bjornfor** [comment]
+1 when my comments have been addressed.

File: pkgs/applications/virtualization/runc/default.nix (hunk 1 above)
Comment (bjornfor) [inline pkgs/applications/virtualization/runc/default.nix]: Working packaging is more important then strictly sticking to released versions. IOW, I was just making a comment. Not a blocker.

**Comment by matejc** [comment]
+1 from me, I have tested this with docker-compose 1.8.0, that was merged few days back, besides the issue that you can not use usernames anymore `docker run -it --rm --user "matejc" -v /etc/passwd:/etc/passwd:ro ...` but you have to use `docker run -it --rm --user 1001 ...` because bound mounts happen after resolving usernames, but that is a docker/runc issue (according to their issue tracker is not going to be fixed)

File: pkgs/applications/virtualization/runc/default.nix (hunk 1 above)
Comment (offlinehacker) [inline pkgs/applications/virtualization/runc/default.nix]: Well apparently stable releases of runc are older than commit, i will replace with commit date.

File: pkgs/development/tools/misc/md2man/default.nix (hunk 2 above)
Comment (offlinehacker) [inline pkgs/development/tools/misc/md2man/default.nix]: tnx, will fix

File: nixos/modules/virtualisation/docker.nix (hunk 3 above)
Comment (offlinehacker) [inline nixos/modules/virtualisation/docker.nix]: Ok, will add in description. It's acually not new, we used to have by default to docker to choose driver, but it was changed later to default option that broke on some filesystems like btrfs, and was not a good choice for some others. Docker has a good priority list which device storage to use.

**Comment by offlinehacker** [comment]
@bjornfor i have addressed issues you mentioned, tnx! @matejc thanks for testing, i think we can live with uid and gid numbers :)

File: nixos/modules/virtualisation/docker.nix [hunk 4]
```diff
@@ -40,13 +40,25 @@ in
       };
     storageDriver =
       mkOption {
-        type = types.enum ["aufs" "btrfs" "devicemapper" "overlay" "zfs"];
-        default = "devicemapper";
+        type = types.nullOr (types.enum ["aufs" "btrfs" "devicemapper" "overlay" "zfs"]);
```
Comment (vdemeester) [inline nixos/modules/virtualisation/docker.nix]: Note that there is a `overlay2` driver with 1.12. The only requirement for the driver is a kernel > 4.0 if I remember correctly.

File: nixos/modules/virtualisation/docker.nix [hunk 5]
```diff
@@ -88,7 +100,12 @@ in
         after = [ "network.target" ] ++ (optional cfg.socketActivation "docker.socket") ;
         requires = optional cfg.socketActivation "docker.socket";
         serviceConfig = {
-          ExecStart = "${pkgs.docker}/bin/docker daemon --group=docker --storage-driver=${cfg.storageDriver} ${optionalString cfg.socketActivation "--host=fd://"} ${cfg.extraOptions}";
+          ExecStart = ''${pkgs.docker}/bin/docker daemon \
```
Comment (vdemeester) [inline nixos/modules/virtualisation/docker.nix]: It's safer to use `dockerd` starting at version 1.12 as `docker daemon` is deprecated 👼

File: nixos/modules/virtualisation/docker.nix (hunk 4 above)
Comment (offlinehacker) [inline nixos/modules/virtualisation/docker.nix]: Ah yes, will add that one to the list.

File: nixos/modules/virtualisation/docker.nix (hunk 5 above)
Comment (offlinehacker) [inline nixos/modules/virtualisation/docker.nix]: Yes, makes sense, will change to dockerd then.

File: nixos/modules/virtualisation/docker.nix (hunk 4 above)
Comment (offlinehacker) [inline nixos/modules/virtualisation/docker.nix]: done

File: nixos/modules/virtualisation/docker.nix (hunk 5 above)
Comment (offlinehacker) [inline nixos/modules/virtualisation/docker.nix]: done

**Comment by offlinehacker** [comment]
Ok, i have additionaly added extra build output called `client`, that has only client binary, without all the dependencies that daemon needs. This is for installations where you only need a client (docker swarm). And some other minor improvements. If you want you can rereview docker expression. Thanks!

**Comment by offlinehacker** [comment]
I've additionally added `docker-client` to all-packages which refers to client build output for those that only need docker client. Some day we should also make docker client build on all other platforms that docker client binary runs (osx).

**Comment by np** [comment]
I'm getting:

```
collision between ‘/nix/store/jp9vg0rirwywbxdig8f7257zy4hi6387-docker-1.12.1-client/bin/docker’ and ‘/nix/store/7jbrm8d5w3j98h89ssfdrvkzipjdnz0p-docker-1.12.1/bin/docker’; use ‘nix-env --set-flag priority NUMBER PKGNAME’ to change the priority of one of the conflicting packages
```

**Comment by offlinehacker** [comment]
I don't know why nix tries to install all outputs, i will remove client output for now...

**Comment by offlinehacker** [comment]
Ok, i have removed client output, now it installs without problems.

**Comment by domenkozar** [comment]
See extraOutputsToInstall attribute.

**Comment by offlinehacker** [comment]
Yes, i've tried, does not help, looks like `buildenv.pl` does not respect that and also current `buildenv.cc`.

**Comment by offlinehacker** [comment]
Cool, let's merge this one, all checks passing, if anyone finds a bug later, i'll fix it :)

**Comment by NeQuissimus** [comment]
Could we cherry-pick this into 16.09? :D

**Comment by vdemeester** [comment]
@offlinehacker @NeQuissimus That would be absolutely awesome if we could :angel:

**Comment by domenkozar** [comment]
Go ahead

File: pkgs/applications/virtualization/containerd/default.nix [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp bin/* $out/bin
+  '';
+
+  preFixup = ''
+    # remove references to go compiler
+    while read file; do
+      sed -ri "s,${go},$(echo "${go}" | sed "s,$NIX_STORE/[^-]*,$NIX_STORE/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee,"),g" $file
```
Comment (zimbatm) [inline pkgs/applications/virtualization/containerd/default.nix]: Is that still necessary when moving the buildInputs to nativeBuildInputs?

**Comment by offlinehacker** [comment]
As far as I know yes, since nix scans all files for files referencing nix
store. The same we do for go packages, but if it's not needed please let me
know.

**Comment by offlinehacker** [comment]
Yes, please someone cherry pick to 16.09 and also please at the same time
new docker compose.

**Comment by vdemeester** [comment]
@offlinehacker I'll take a look at the `docker-compose` one and maybe `go-1.7` if it's ok (/cc @domenkozar)

**Comment by NeQuissimus** [comment]
Is `docker-compose` 1.8.0 not already in 16.09?!

**Comment by vdemeester** [comment]
@NeQuissimus yeah, did not checked before writing that, it's already in 16.09 :wink:

**Pull request merged**